### ISSUE TITLE
Switch to persistent Cart implementation

### ISSUE THREAD

**Issue by odrotbohm** [st-tu-dresden/salespoint#217]
Currently we keep `Cart` instances in the session which is fine, but it also has a few downsides:

- students have to understand the concept of a session, i.e. keeping client specific state between requests
- the former is bad practice anyway (keeping client specific state, not being a student)
- `Cart` instances are lost between restarts even with persistent storage enabled

We should investigate turning that session state into application state by:

- turning `Cart` into an entity just like the others
- find a mechanism to keep it "fresh", i.e. make sure that the we make sure that using the application marks the `Cart` as in use if it is connected to a user account
- some cleanup mechanism that removes unused `Cart` instances not tied to users.

Not sure we actually end up implementing it as we probably need to implement quite a few components manually that could complicate the picture.

**Comment by MarcKandler** [comment]
I actually like the fact that the `Cart` is stored in the session. This is a nice example to illustrate exactly that not everything does need to be stored persistently (and how it can be done).
The current implementation requires the students to develop different trains of thought, which I like to ask for during our weekly meetings.

In the broad context of an example application, I would keep it as it is (maybe with a note stating the mentioned downsides).

**Comment by odrotbohm** [comment]
The thing is that the session is something that should completely avoided as it makes an application unRESTful and less scalable (see Fielding [3.4.3](https://www.ics.uci.edu/~fielding/pubs/dissertation/net_arch_styles.htm#sec_3_4_3) and [5.1.3](https://www.ics.uci.edu/~fielding/pubs/dissertation/rest_arch_style.htm#sec_5_1_3)).

Even more important in our specific scenario: it requires the students to understand a completely new concept that's never actually used anywhere else in Salespoint. The only concept sort of similar is authentication but that's nicely hidden behind the ability to use `UserAccount` with a `@LoggedIn` annotation.

If you look at `OrderController`, the implementation of creating and using the `Cart` requires implementing a couple of extra methods, use a class-level attribute all of which are not really obvious why they need to be there and how they work together. I'd argue, this leads more to a copy-and-paste-to-get-things-working approach rather than really understanding what's going on.

**Comment by martinmo** [comment]
> Even more important in our specific scenario: it requires the students to understand a completely new concept that's never actually used anywhere else in Salespoint.

That's true, but aren't sessions at least used widely in day-to-day web development and it's worth learning? I always thought that a shopping cart was the prime example for a session attribute :-)

> The only concept sort of similar is authentication but that's nicely hidden behind the ability to use UserAccount with a @LoggedIn annotation.
> If you look at OrderController, the implementation of creating and using the Cart requires implementing a couple of extra methods, use a class-level attribute all of which are not really obvious why they need to be there and how they work together.

To make it simpler to understand, reduce boilerplate and reuse existing concepts: What about making `Cart` a bean with `@SessionScope` (https://docs.spring.io/spring-framework/docs/current/spring-framework-reference/core.html#beans-factory-scopes-session)?

> The thing is that the session is something that should completely avoided as it makes an application unRESTful and less scalable (see Fielding 3.4.3 and 5.1.3).

Nevertheless, with the proposed change, we would not meet Fieldings criteria because the state is still stored on the server between requests. It just uses a different storage (not in-memory as part of the `HttpSession`, but in the database).

**Comment by odrotbohm** [comment]
> That's true, but aren't sessions at least used widely in day-to-day web development and it's worth learning?

I'd argue that there's way more important things to learn in this lab than a dumbed down variant of user-specific application state. The focus is on designing and architecting a decent system. It shouldn't be about a hardly used feature of an implementation technology that exists for legacy reasons.

> I always thought that a shopping cart was the prime example for a session attribute :-)

It's the same kind of a prime example for session usage as transferring money between accounts is for a transaction: it's not. Neither does transferring money actually work with a database transaction, nor are carts implemented using sessions in the real world.

> To make it simpler to understand, reduce boilerplate and reuse existing concepts: What about making Cart a bean with @SessionScope.

That'd solve the code clutter issue. An issue that'd still to be solved is that carts would die between application restarts even with durable persistence enabled.

It would then still also mean we'd have to explain the concept of session scope which brings us back to square one. We'd also still need some glue code to avoid the `Cart` instances to be injected into and to be held in a field in the controller, which then requires me to explain, why we recommend to keep Spring components stateless. However, registering a session scoped bean for carts might actually help in the actual implementation.

> Nevertheless, with the proposed change, we would not meet Fieldings criteria because the state is still stored on the server between requests. It just uses a different storage (not in-memory as part of the `HttpSession`, but in the database).

Here's the relevant part from the section:

> The client-stateless-server style derives from client-server with the additional constraint that no *session state* is allowed on the server component.

I.e. REST is completely fine with application state. In fact, it wouldn't actually work without it. No web application would ever work without it unless it's computations based on the request input only. Session state is to be avoided, so that you can horizontally scale, i.e. the clients don't have to worry about which particular server of a cluster they talk to. With the session in use, they have to talk to the same instance for every request.
